### PR TITLE
sql: add typecasts to expressions shown in pg_catalog

### LINKED ISSUE #53942: sql: pg_catalog should add typecasts when showing formatted SQL expressions
This was broken accidentally in PR #53692 .

When showing DEFAULT expressions and CHECK constraint expressions in pg_catalog, the result is supposed to include a typecast for strings, enums, and other quoted types.

For example, see the following Postgres behavior.

```
rafiss@127:postgres> CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi');
CREATE TYPE

rafiss@127:postgres> CREATE TABLE enums (
    x INT,
    y greeting DEFAULT 'hello',
    z greeting CHECK (z != 'hi'));
CREATE TABLE

rafiss@127:postgres> SELECT
    pg_get_expr(d.adbin, d.adrelid)
  FROM
    pg_attribute AS a
    LEFT JOIN pg_attrdef AS d ON a.attrelid = d.adrelid AND a.attnum = d.adnum
  WHERE
    a.attrelid = 'enums'::REGCLASS
    AND attname = 'y';
+-------------------+
| pg_get_expr       |
|-------------------|
| 'hello'::greeting |
+-------------------+

rafiss@127:postgres> select pg_get_constraintdef(con.oid) FROM pg_catalog.pg_constraint con where con.conrelid = 'enums'::regclass;;
+-------------------------------+
| pg_get_constraintdef          |
|-------------------------------|
| CHECK ((z <> 'hi'::greeting)) |
+-------------------------------+
```

CockroachDB stopped including the typecasts.

This is breaking parsing logic in ActiveRecord. See the [following code](https://github.com/rails/rails/blob/7da82b0fc32a3257d9bc44605783d8aa1a7474a1/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L590):

```
        def extract_value_from_default(default)
          case default
            # Quoted types
          when /\A[\(B]?'(.*)'.*::"?([\w. ]+)"?(?:\[\])?\z/m
            # The default 'now'::date is CURRENT_DATE
            if $1 == "now".freeze && $2 == "date".freeze
              nil
            else
              $1.gsub("''".freeze, "'".freeze)
            end
            # Boolean types
          when "true".freeze, "false".freeze
            default
            # Numeric types
          when /\A\(?(-?\d+(\.\d*)?)\)?(::bigint)?\z/
            $1
            # Object identifier types
          when /\A-?\d+\z/
            $1
          else
            # Anything else is blank, some user type, or some function
            # and we can't know the value of that, so return nil.
            nil
          end
        end
```

There is logic in [AnnotateTypeExpr.format()](https://github.com/cockroachdb/cockroach/blob/91c7ae24d485695210cd6f9d06d02a9982d7bf9c/pkg/sql/sem/tree/expr.go#L1654) to show the typecasts, but it was recently broken.

### PR DESCRIPTION
**Pull request by rafiss**
fixes #53942 

also includes three small cleanup commits:
colexec: reduce code duplication in mergejoiner
descpb: fix comment about formatting expressions
sql: consolidate FormatExprForDisplay functions

pg_attrdef.adbin (column defaults) and pg_constraint.condef (column
constraints) are supposed to format the expression using typecasts for
non-numeric constants. This was previously being done for string
constants in pg_attrdef, but a recent change broke that functionality.

Release note (sql change): SQL expressions that are shown in pg_catalog
columns and related functions now will be formatted with a typecast for
non-numeric constants.

Release justification: low-risk change that improves compatibility, and
also fixes a regression that started in this release.

### CODE REVIEW

File: pkg/sql/schemaexpr/expr.go [hunk 1]
```diff
@@ -136,6 +136,21 @@ func FormatExprForDisplayWithoutTypeAnnotations(
 	return tree.AsString(expr), nil
 }
 
+// FormatExprForDisplayInPGCatalog formats a schema expression string for
+// display, similar to FormatExprForDisplay, but does not add type annotations
+// and does add typecasts for non-numeric constants. This is the format that
+// should be used for displaying expressions in pg_catalog, in order to achieve
+// compatibility with the pg_catalog from PostgreSQL.
+func FormatExprForDisplayInPGCatalog(
```
Comment (mgartner) [inline pkg/sql/schemaexpr/expr.go:144]: [nit] Create a format option enum and combine these three `FormatExprForDisplay...` functions into one. They're all very similar.

File: pkg/sql/sem/tree/format.go [hunk 2]
```diff
@@ -365,7 +366,7 @@ func (ctx *FmtCtx) FormatNode(n NodeFormatter) {
 			ctx.WriteByte(')')
 		}
 	}
-	if f.HasFlags(fmtDisambiguateDatumTypes) {
+	if f.HasFlags(fmtDisambiguateDatumTypes) || f.HasFlags(FmtPGCatalog) {
```
Comment (mgartner) [inline pkg/sql/sem/tree/format.go:369]: [nit] `if f.HasFlags(fmtDisambiguateDatumTypes | FmtPGCatalog) {`

**Review by mgartner — APPROVED** [review]
LGTM, just 2 optional changes. Happy to take another look if you decide to make the changes.

File: pkg/sql/schemaexpr/expr.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/schemaexpr/expr.go:144]: good call

File: pkg/sql/sem/tree/format.go (hunk 2 above)
Comment (rafiss) [reply pkg/sql/sem/tree/format.go:369]: done

**Comment by rafiss** [comment]
This is ready for another look

File: pkg/sql/schemaexpr/expr.go [hunk 3]
```diff
@@ -111,29 +111,21 @@ func ExtractColumnIDs(desc catalog.TableDescriptor, rootExpr tree.Expr) (TableCo
 	return colIDs, err
 }
 
-// FormatExprForDisplay formats a schema expression string for display by adding
-// type annotations and resolving user defined types.
+// FormatExprForDisplayInPGCatalog formats a schema expression string for
```
Comment (mgartner) [inline pkg/sql/schemaexpr/expr.go:114]: -> `FormatExprForDisplay`

File: pkg/sql/pg_catalog.go [hunk 4]
```diff
@@ -414,19 +413,11 @@ CREATE TABLE pg_catalog.pg_attrdef (
 				// pg_attrdef only expects rows for columns with default values.
 				return nil
 			}
-			var defSrc *tree.DString
-			expr, err := parser.ParseExpr(*column.DefaultExpr)
+			displayExpr, err := schemaexpr.FormatExprForDisplayInPGCatalog(ctx, table, *column.DefaultExpr, &p.semaCtx)
```
Comment (mgartner) [inline pkg/sql/pg_catalog.go:416]: -> `FormatExprForDisplay(..., tree.FmtPGCatalog)`

File: pkg/sql/pg_catalog.go [hunk 5]
```diff
@@ -978,7 +969,7 @@ func populateTableConstraints(
 			if conkey, err = colIDArrayToDatum(con.CheckConstraint.ColumnIDs); err != nil {
 				return err
 			}
-			displayExpr, err := schemaexpr.FormatExprForDisplayWithoutTypeAnnotations(ctx, table, con.Details, &p.semaCtx)
+			displayExpr, err := schemaexpr.FormatExprForDisplayInPGCatalog(ctx, table, con.Details, &p.semaCtx)
```
Comment (mgartner) [inline pkg/sql/pg_catalog.go:972]: -> `FormatExprForDisplay(..., tree.FmtPGCatalog)`

**Review by mgartner — COMMENTED** [review]
Nice! Looks great, you just missed 3 things.

**Comment by rafiss** [comment]
Thanks! seems like I rushed too much

**Review by mgartner — APPROVED** [review]
```
 _     ____ _____ __  __ 
| |   / ___|_   _|  \/  |
| |  | |  _  | | | |\/| |
| |__| |_| | | | | |  | |
|_____\____| |_| |_|  |_|
```

File: pkg/sql/sem/tree/format.go (hunk 2 above)
Comment (rafiss) [reply pkg/sql/sem/tree/format.go]: Turns out `HasFlags` checks if _all_ of the flags are set, so I wrote a new `HasAnyFlags` function and used that.

**Comment by rafiss** [comment]
Turns out `HasFlags` checks if all of the flags are set, so I wrote a new `HasAnyFlags` function and used that.

**Comment by rafiss** [comment]
bors r=mgartner

**Pull request merged**